util/retrieve-cluster-logs.py

```diff
@@ -0,0 +1,213 @@
+import gzip
+import json
+import logging
+import os
+import sys
+import tarfile
+import tempfile
+from datetime import datetime
+
+import argparse
+import boto3
+
+logging.basicConfig(level=logging.INFO)
+LOGGER = logging.getLogger(__file__)
+
+
+def err_and_exit(message):
+    """Log the given error message and exit nonzero."""
+    LOGGER.error(message)
+    sys.exit(1)
+
+
+def get_default_bucket_prefix(cluster_name):
+    """Get default keypath to store logs under in an s3 bucket for the given cluster."""
+    return "{cluster_name}-logs-{end_time}".format(
+        cluster_name=cluster_name, end_time=datetime.now().strftime("%Y%m%d%H%M%S")
+    )
+
+
+def get_log_group(args):
+    """Get log group for args.cluster."""
+    log_group_name = "/aws/parallelcluster/{}".format(args.cluster)
```

Comment:
We should have a global variable defined on top of the module for the group name path or use a global variable from the source code

Reply:
What's the motivation for making it a global?

Is your concern that this function is too expensive to call every time the log group name is needed? I agree that it is. This function is only called once when parsing the command line. It has to be done then because it requires the cluster's name as input. The returned value is saved in the args object that's passed around the rest of the script, so the function is not called again.

Reply:
The "/aws/parallelcluster/" path is something like a configuration, that must correspond to the one used in the cloudwatch resource creation and in the rest of the code, so I think we should have a single place on which it is defined, if in future we want to change it.

If we want to keep this script separate from the rest of the code I'd move the definition of the log group name on top, to be clear when you open the file what are the "configuration" parameters.

What do you think?

Reply:
The prefix is used in four places within the code base:
* cookbook
* cfn template
* integration tests
* this script

I don't know of a way to share with the first two items in this list a definition for this prefix. If you know of a way I could do that I'd be open to trying it out.

For sharing the definition between the last two items, my approach would be to create a subpackage `pcluster.cloudwatch_logs` with a `utils` module that contains a function that wraps the prefix. I've made this change and am pushing it now. Let me know what you think.

Reply:
It is good for me. 👍 
Alas we don't have a way (yet?) to share information between CFN, cookbook and python.

For the mapping file I created a test_source_consistency.py unit test that verifies if the default set in CFN are the same used for the defaults in the mappings.py, but cookbook files are totally separated.

Reply:
Ah, I didn't think about the consistency tests. I'll add one.

Reply:
Actually after looking at those tests I don't think it goes well there since the name is not passed as a parameter. I'll omit a test for now.